File: sccs/sccs.py
import load_data

G = load_data.load_data()
NODES = 875715

explored = [False] * NODES
leader = [False] * NODES
finishing_times = [False] * NODES

t = 0
run_leader = None

nodes = [[] for _ in range(NODES)]
nodes_rev = [[] for _ in range(NODES)]

for edge in G:
    [i, j] = edge
    nodes[i].append(j)
    nodes_rev[j].append(i)

def DFS_Loop(order, nodes):
    global t, run_leader, explored

    for i in order:
        if not explored[i]:
            run_leader = i
            DFS_iterative(nodes, i)


def DFS_iterative(nodes, v):
    global explored, t, finishing_times, leader, run_leader
    stack = [v]

    while len(stack):
        v = stack.pop()

        if not explored[v]:
            explored[v] = True
            leader[v] = run_leader
            stack.append(v)
            for w in nodes[v]:
                if not explored[w]:
                    stack.append(w)
        else:
            if not finishing_times[v]:
                finishing_times[v] = t
                t = t + 1

DFS_Loop(range(1, len(nodes)), nodes_rev)

finishing_times[0] = NODES - 1
nodes_f = [None] * NODES

nodes_order = list(range(len(nodes)))
for i, f in enumerate(finishing_times):
    nodes_f[finishing_times[i]] = nodes_order[i]
nodes_f = list(reversed(nodes_f))

explored = [False] * NODES
leader = [False] * NODES
finishing_times = [False] * NODES

t = 0
run_leader = None
DFS_Loop(nodes_f, nodes)

leaders = {}

for i in leader:
    if leaders.get(i):
        leaders[i] = leaders[i] + 1
    else:
        if i:
            leaders[i] = 1

print(sorted(list(leaders.values()), reverse=True)[:10])

File: sccs/sccs.cpp
#include <iostream>
#include <fstream>
#include <string>
#include <vector>

using namespace std;

int f[875714] = {0};
int explored[875714] = {0};
int leader[875714] = {0};

int t = 0;
int s = 0;

int* getEdges(string text) {
    int separation = text.find(" ");
    int start = stoi(text.substr(0, separation));
    int end = stoi(text.substr(separation, text.length()));
    int *result = new int[2]{start, end};
    return result;
}

vector<vector<int>> changeOrder(vector<vector<int>> nodes, vector<int> newOrder) {
    vector<vector<int>> result;
    for (int i = 0; i < newOrder.size(); i++) {
        result.push_back(nodes[newOrder[i]]);
    }
    return result;
}

void DFS(vector<vector<int>> Graph, int i) {
    ::explored[i] = 1;
    ::leader[i] = ::s;

    for (int j: Graph[i]) {
        if (!::explored[j]) {
            DFS(Graph, j);
        }
    }

    ::t = ::t + 1;
    ::f[i] = ::t;
}

void DFS_Loop(vector<vector<int>> Graph) {
    for (int i = 0; i < Graph.size(); i++) {
        if (!::explored[i]) {
            ::s = i;
            DFS(Graph, i);
        }
    }
}

int main() {
    cout << "Hola" << endl;

    vector<vector<int>> G;
    vector<int> node0 = {0};

    G.push_back(node0);

    for (int i = 0; i < 875714; i++) {
        vector<int> node;
        G.push_back(node);
    }


    string text;
    ifstream EdgesFile("sccs.txt");

    int myArray[10] = { 0 };


    while (getline (EdgesFile, text)) {
        int *r = getEdges(text);
        G[r[0]].push_back(r[1]);
    }
    EdgesFile.close();

    for (auto i: G[1]) {
        cout << i << ' ';
    }

    DFS_Loop(G);

    vector<int> f_vec(begin(f), end(f));
    vector<vector<int>> G_ = changeOrder(G, f_vec);
    // nodes_f = [value for (value, index) in sorted(zip(nodes, f), reverse=True)]

    for (int i = 0; i < 875714; i++) {
        cout << leader[i] << endl;
    }

    // for (int i = 0; i < 875714; i++) {
    //     f[i] = 0;
    //     explored[i] = 0;
    //     leader[i] = 0;
    // }

    // DFS_Loop(G_);

    // for (int i = 0; i < 875714; i++) {
    //     cout << leader[i] << endl;
    // }

    // r = {}

    // for i in f:
    //     if r.get(i):
    //         r[i] = r[i] + 1
    //     else:
    //         r[i] = 1
    // print('hola')
    // print(sorted(list(r.values()), reverse=True))
}
